src/pools/D3MPoolBase.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021-2022 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+interface TokenLike {
+    function approve(address, uint256) external returns (bool);
+    function transfer(address, uint256) external returns (bool);
+    function transferFrom(address, address, uint256) external returns (bool);
+    function balanceOf(address) external view returns (uint256);
+}
+
+interface CanLike {
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface d3mHubLike {
+    function vat() external view returns (address);
+}
+
+abstract contract D3MPoolBase {
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "D3MPoolBase/not-authorized");
+        _;
+    }
+
+    TokenLike   public immutable asset; // Dai
+    address     public immutable hub;
+
+    uint256     public           live = 1;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Cage();
+
+    constructor(address hub_, address dai_) internal {
+        asset = TokenLike(dai_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        hub = hub_;
+        wards[hub_] = 1;
+        emit Rely(hub_);
+        CanLike(d3mHubLike(hub_).vat()).hope(hub_);
+    }
+
+    function validTarget() external view virtual returns (bool);
+
+    function deposit(uint256 amt) external virtual;
+
+    function withdraw(uint256 amt) external virtual;
+
+    function transfer(address dst, uint256 amt) external virtual returns (bool);
+
+    function transferAll(address dst) external virtual returns (bool);
+
+    function accrueIfNeeded() external virtual;
+
+    function assetBalance() external view virtual returns (uint256);
+
+    function maxWithdraw() external view virtual returns (uint256);
+
```

Comment:
Need to add `maxDeposit`. This is not relevant for Aave or Compound, but it is relevant for Maple and TrueFi. Should just return `type(uint256).max` for Aave/Compound.

Reply:
which contract would call that function and where?

Reply:
I assume that would provide an external upper bound on our wind function?

Reply:
Yes what @iamchrissmith said.

Reply:
@hexonaut is maxDeposit the 
1. maximum amount that can be deposited at once
2. the max size any address can have deposited
3. the max number of total assets that can be deposited
4. ???

Reply:
let me know what you think of this implmentation: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/c341efa87ad65817e4da9017a15ee71c4e4c6f1a

There was also the case where maxDeposit is greater than our current position so we need to unwind.  I believe I captured/tested both

Reply:
I mentioned in the new code, but looking at the implementation it seems that somehow could be merged with what `ilks[ilk_].plan.getTargetAssets(currentAssets)` returns as it is immediately overwritten.

Reply:
@hexonaut how do you feel about removing this from the pool interface and limiting in the hub in favor of putting the logic in the plans that need it?

Reply:
Chris, I think actually is fine how you originally did. It makes more sense to have better separation of concerns instead of gas savings for this specific module.
On another topic, I have my doubts about `maxDeposit` provoking unwinding. Should we actually do it?
I mean, a downsize of the deposit limit from their side, shouldn't in reality force us to unwind, like our vault holders are not forced to unwind if the debt ceiling is reduced.
It's different from reducing our debt ceilings because there we are the ones that want to reduce the risk.
IMO `maxDeposit` should be only for `wind`.

Reply:
Yes how you added it is good Chris (it's not quite correct though). In general let's not worry about gas too much as this is infrequently called for large sums of DAI. Aave one for example is called once every few days for adding/removing tens of million of DAI at once.

`maxDeposit` should return the amount that can be added to the pool by anyone regardless of how much is in there. As such it should limit the amount to deposit and not the total target assets as the commit above does. `maxDeposit` should never be the thing that triggers an unwind. Think of it like how much debt ceiling room is available. It should be used in the `wind()` upper limit constraint (same as we do for available debt ceiling on our end).

Reply:
Well in this version we are actually unwinding if for any reason we end up above the debt ceiling.
I think in the case of our debt ceiling makes sense as it is our governance that reduces it and there is a reason to do that.
In case of `maxDeposit` I agree it should only be for `wind`ing.

Reply:
ahh yes I see.  That is how I originally had it but misread 4626 to be total deposits not per `deposit` call.  Updated: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/d86843b9055bfcab66d129f905723bfad98dc899

Reply:
Looks good now.